packages/metrics/src/DimensionsStore.ts

```diff
@@ -0,0 +1,67 @@
+import { InvokeStore } from '@aws/lambda-invoke-store';
+import type { Dimensions } from './types/Metrics.js';
+
+/**
+ * Manages storage of metrics dimensions with automatic context detection.
+ *
+ * This class abstracts the storage mechanism for metrics, automatically
+ * choosing between AsyncLocalStorage (when in async context) and a fallback
+ * object (when outside async context). The decision is made at runtime on
+ * every method call to support Lambda's transition to async contexts.
+ */
+class DimensionsStore {
+  #fallbackDimensions: Dimensions = {};
+  #fallbackDimensionSets: Dimensions[] = [];
+
+  #getDimensions(): Dimensions {
+    if (InvokeStore.getContext() !== undefined) {
+      let stored = InvokeStore.get('dimensions') as Dimensions | undefined;
+      if (stored == null) {
+        stored = {};
+        InvokeStore.set('dimensions', stored);
+      }
+      return stored;
+    }
+    return this.#fallbackDimensions;
+  }
+
+  #getDimensionSets(): Dimensions[] {
+    if (InvokeStore.getContext() !== undefined) {
+      let stored = InvokeStore.get('dimensionSets') as Dimensions[] | undefined;
+      if (stored == null) {
+        stored = [];
+        InvokeStore.set('dimensionSets', stored);
+      }
+      return stored;
+    }
+    return this.#fallbackDimensionSets;
+  }
+
+  addDimension(name: string, value: string): void {
+    this.#getDimensions()[name] = value;
+  }
+
+  addDimensionSet(dimensionSet: Dimensions): void {
+    this.#getDimensionSets().push({ ...dimensionSet });
+  }
+
+  getDimensions(): Dimensions {
+    return { ...this.#getDimensions() };
+  }
+
+  getDimensionSets(): Dimensions[] {
```

Comment:
I have possibly over-indexed on immutability here but my reasoning is that I don't want it to be possible for some to get a reference to the underlying object in the store. Any mutations of this stores should always be done through the interface.